### ISSUE TITLE
GDAX - "Price too precise" on latest `stable`

### ISSUE THREAD

**Issue by GuyPaddock** [askmike/gekko#992]
* **I'm submitting a ...**
[x] bug report
[ ] feature request
[ ] question about the decisions made in the repository

* **Action taken** (what you did)
Left a MACD BTC/USD trading strategy running on GDAX throughout the day via CLI.

* **Expected result** (what you hoped would happen)
BUY and SELL orders would be processed without errors.

* **Actual result** (unexpected outcome)
BUY orders that are too precise fail, as follows:
```
2017-08-17 10:04:04 (DEBUG):    Processing 8 new trades. From 2017-08-17 14:03:50 UTC to 2017-08-17 14:04:01 UTC. (a few seconds)
2017-08-17 10:04:04 (DEBUG):    calculated MACD properties for candle:
2017-08-17 10:04:04 (DEBUG):             short: 4449.26238286
2017-08-17 10:04:04 (DEBUG):             long: 4431.91163462
2017-08-17 10:04:04 (DEBUG):             macd: 17.35074824
2017-08-17 10:04:04 (DEBUG):             signal: 17.08544991
2017-08-17 10:04:04 (DEBUG):             macdiff: 0.26529833
2017-08-17 10:04:04 (DEBUG):    In uptrend since 1 candle(s)

2017-08-17 10:04:04 (INFO):     We have new trading advice!
2017-08-17 10:04:04 (INFO):              Position: long
2017-08-17 10:04:04 (INFO):              Market price: 4467.99
2017-08-17 10:04:04 (INFO):              Based on market time: 2017-08-17 14:03:00

2017-08-17 10:04:04 (INFO):     Trader Received advice to go long. Buying  BTC
2017-08-17 10:04:04 (INFO):     Attempting to BUY 0.02090085414447822 BTC at GDAX price: 4467.97999999
2017-08-17 10:04:05 (ERROR):    Error buying at GDAX: { message: 'price too precise (4467.97999999)' }
2017-08-17 10:04:05 (DEBUG):    GDAX returned an error, retrying..
2017-08-17 10:04:16 (ERROR):    Error buying at GDAX: { message: 'price too precise (4467.97999999)' }
2017-08-17 10:04:16 (DEBUG):    GDAX returned an error, retrying..
2017-08-17 10:04:24 (DEBUG):    Requested BTC/USD trade data from GDAX ...
2017-08-17 10:04:24 (DEBUG):    Processing 9 new trades. From 2017-08-17 14:04:10 UTC to 2017-08-17 14:04:24 UTC. (a few seconds)
2017-08-17 10:04:31 (ERROR):    Error buying at GDAX: { message: 'price too precise (4467.97999999)' }
2017-08-17 10:04:31 (DEBUG):    GDAX returned an error, retrying..
2017-08-17 10:04:41 (ERROR):    Error buying at GDAX: { message: 'price too precise (4467.97999999)' }
2017-08-17 10:04:41 (DEBUG):    GDAX returned an error, retrying..
2017-08-17 10:04:44 (DEBUG):    Requested BTC/USD trade data from GDAX ...
```

* **Other information**
- I am on the `stable` branch at 1c0700d312e3f6bc6e4fb9f4d1fdcda5d3392736.
- It looks like even after #922, USD prices are still too precise.

**Comment by askmike** [comment]
ping @damianleung

**Comment by wingleung** [comment]
@askmike @GuyPaddock just got back from abroad, had a quick look, see bugfix on #1002 
Currently running on my setup for testing on ETH-EUR.

I don't know where the USD difference came up though. On EUR the price is always 2 decimals and `this.getMaxDecimalsNumber(123.12)` always returns "123.12".

And in @GuyPaddock's log the market price is 4467.99, so `this.getMaxDecimalsNumber(4467.99)` should return "4467.99".

Or do I have to take these parameters in consideration as well?
`short: 4449.26238286`
`long: 4431.91163462`

**Comment by askmike** [comment]
> Or do I have to take these parameters in consideration as well?

No these are never passed to GDAX.

So the price was 4467.99 but Gekko tried to pass `4467.97999999`, which has exactly 8 decimals. Maybe we should force the price to 2 for fiat markets and 5 for cryptomarkets?

**Comment by wingleung** [comment]
@askmike where is 5 coming from? wasn't it 8? For now I've committed 8 decimals for BTC, 2 decimals for all other currencies, see #1002 If it has to be 5 I'll change it

**Comment by askmike** [comment]
I commented in #1002!

**Comment by wingleung** [comment]
@askmike I removed the deductFee() already, or did you mean to comment something else?

As for the pricing decimals...
`2017-08-17 10:04:04 (INFO):     Attempting to BUY 0.02090085414447822 BTC at GDAX price: 4467.97999999`
this line comes from https://github.com/askmike/gekko/blob/eec954a6d9eebe720530fe77fff2d2d89fe890f5/plugins/trader/portfolioManager.js#L215-L223

I think that might be the issue and could introduce similar issues on other exchanges.
Unless we decide to trim down to the appropriate decimals in the exchange layer, which means we have to check for flat/crypto markets in every exchange and trim accordingly.
But I think portfoliomanager is a better place to do this.
https://github.com/askmike/gekko/blob/eec954a6d9eebe720530fe77fff2d2d89fe890f5/plugins/trader/portfolioManager.js#L158-L167

Fairly new to crypto trading so I could be looking at this in a wrong way..

**Comment by askmike** [comment]
@GuyPaddock it should be fixed now (thanks to @damianleung), can you try again?

**Comment by GuyPaddock** [comment]
I'll give it a go! Thanks, guys!

**Comment by bettimms** [comment]
Recently got the same error trying to trade ZRX/BTC on gdax.

**Comment by askmike** [comment]
@bettimms can you try latest develop? Some changes were pushed here: #2597.

If that doesn't solve it please open a new issue.
